### DISCUSSION TITLE
Check if Vector_3 is assigned

### DISCUSSION THREAD

**Discussion by petrasvestartas** [CGAL/cgal#6045] [Q&A]
Hi,

I have a simple question.
How can I check if a `CGAL_Kernel::Vector_3 `is assigned?

For example I declare variable (first case)
`CGAL_Kernel::Vector_3 v;`
and like this (second case):
`CGAL_Kernel::Vector_3 v(0,0,1);`

Is there any way to know if a user initialized the object or not?
In the first case the return value would be false, in the second true.

Is there a shorter way that doing this?
`bool isValid = ((v.x()+v.y()+v.z())<0.0001); `

**Comment by petrasvestartas** [comment]
Anyone?

**Comment by afabri** [comment]
You cannot check.  We had a discussion (or even github issue?) about initializing with  NaN to make debugging easier, but  didnt conclude.

**Comment by lrineau** [comment]
There is nothing general, in C++, that can tell you if a variable was initialized or not.

If that is important for you, then use special class templates `std::optional` or `boost::optional`: the value-initialized (default-constructed) `std::optional` has the "empty" state.

**Comment by petrasvestartas** [comment]
Thank you @afabri  and @lrineau 

I will use a sum of coordinates as a check, it works fine for my current case, as they are initial all zeros.

Thank you for your response.

**Reply by lrineau** [reply]
Note that the vector with coordinates (1, 0, -1) has the sum of coordinates equal to 0, as well as (0, 1,-1) and (1, - 1, 0). In fact, there is a full set of vectors like that: any combination of those three vectors, and that is the vector plane of equation x+y+z=0. You should use the equation  `v==CGAL::ORIGIN` instead, so that only the null vector is detected by the criterion.

**Comment by petrasvestartas** [comment]
@lrineau Thank you for pointing this out, now I understand why I had plenty of random errors that I did not understood.
Your suggestion gives error because `v==CGAL::ORIGIN` CGAL::ORIGIN is of a type of origin, not a vector (type is not the same) .
Would it work if I use `bool is_set = v != CGAL_Vector(0,0,0)`?

Another question I have is also a beginner level. 
How can I get a pointer of a vector?
If I write `auto ptr = &v;`
I see that the auto type is again CGAL_Vector, not an address of a vector.

**Reply by lrineau** [reply]
Sorry, for a vectors that is `v==CGAL::NULL_VECTOR`. ORIGIN is for points.

**Reply by petrasvestartas** [reply]
@lrineau Thanks this works.

How about the second part of the question about pointers?

would it be enough to write `CGAL_Vector* pointer = &v`;

**Reply by petrasvestartas** [reply]
It was a silly question, I know the answer.

**Comment by petrasvestartas** [comment]
@lrineau 

I checked again this:
```
CGAL_Vector dir;//not set
bool dirSet = dir != CGAL::NULL_VECTOR;
```

And this produces true;

**Comment by lrineau** [comment]
There is nothing that guarantees that an non-initialized CGAL object will have its coordinates zeroed, and that is for efficiency reasons: the default-constructor leaves the coordinates uninitialized.

As I said, if it really matters for you to detect when a variable is left uninitialized, then you need to:
  - either manually initialize yourself all the values to a sentinel value (using a Nan, `std::numeric_limits<double>::quiet_NaN()` for example, for all three coordinates),
  - or use a class template like `std::optional<CGAL_Vector>` or `boost::optional<CGAL_Vector>` so that you can easily detect the "empty state" of a variable.
